stage1/init/common/mount.go

```diff
@@ -68,131 +69,241 @@ func (m *Mount) Source(podRoot string) string {
 	case "empty":
 		return filepath.Join(common.SharedVolumesPath(podRoot), m.Volume.Name.String())
 	}
-	return "" // We validate in GenerateMounts that it's valid
+	return "" // We validate in RuntimeMounts that it's valid
 }
 
-// GenerateMounts maps MountPoint paths to volumes, returning a list of mounts,
-// each with a parameter indicating if it's an implicit empty volume from a
-// Docker image.
-func GenerateMounts(ra *schema.RuntimeApp, podVolumes []types.Volume, convertedFromDocker bool) ([]Mount, error) {
-	app := ra.App
+// RuntimeMounts represents a mount points generator of the runtime
+// application.
+type RuntimeMounts struct {
+	// Application is the runtime application to generate mount points.
+	Application *schema.RuntimeApp
 
-	var genMnts []Mount
+	// Volumes specifies a list of volumes of the application.
+	Volumes []types.Volume
 
-	vols := make(map[types.ACName]types.Volume)
-	for _, v := range podVolumes {
-		vols[v.Name] = v
+	// DockerImplicit indicates if the resulting mount points should be
+	// marked as implicit empty volumes from a Docker image.
+	DockerImplicit bool
+
+	// processed defines a set of processed mount points, so we don't
+	// have to walk though them multiple times.
+	processed map[string]struct{}
+
+	// The mappings of the POD volumes and the application mount points
+	// for the fast access.
+	volumes volumesMap
+	mounts  mountsMap
+}
+
+// NewRuntimeMounts creates a new instance of the RuntimeMounts with a
+// attributes retrieved from the given POD configuration.
+func NewRuntimeMounts(ra *schema.RuntimeApp, pod *stage1types.Pod) *RuntimeMounts {
+	return &RuntimeMounts{
+		Application: ra,
+		Volumes:     pod.Manifest.Volumes,
+		DockerImplicit: ConvertedFromDocker(
+			pod.Images[ra.Name.String()],
+		),
 	}
+}
 
-	// RuntimeApps have mounts, whereas Apps have mountPoints. mountPoints are partially for
-	// Docker compat; since apps can declare mountpoints. However, if we just run with rkt run,
-	// then we'll only have a Mount and no corresponding MountPoint.
-	// Furthermore, Mounts can have embedded volumes in the case of the CRI.
-	// So, we generate a pile of Mounts and their corresponding Volume
+// volumesMap is a mapping of volume name to volume details.
+type volumesMap map[types.ACName]types.Volume
 
-	// Map of hostpath -> Mount
-	mnts := make(map[string]schema.Mount)
+// mountsMap is a mapping of mount point path to mount point details.
```

Comment:
Ah, I see. the fact that it's a `types.ACName` is a bit confusing, as UNIX paths can have more characters than an ACName.